pkg/agents/audio/speech.go

```diff
@@ -0,0 +1,93 @@
+package audio
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+
+	"github.com/acorn-io/z"
+	cclient "github.com/gptscript-ai/clicky-chats/pkg/client"
+	"github.com/gptscript-ai/clicky-chats/pkg/db"
+	"gorm.io/gorm"
+)
+
+func (a *agent) runSpeech(ctx context.Context) error {
+	slog.Debug("checking for an transcription request to process")
+	var (
+		speechRequest = new(db.CreateSpeechRequest)
+		gdb           = a.db.WithContext(ctx)
+	)
+	if err := db.Dequeue(gdb, speechRequest, a.id); err != nil {
+		return err
+	}
+
+	l := slog.With("type", "speech", "id", speechRequest.ID)
+	l.Debug("processing request")
+
+	data, err := json.Marshal(speechRequest.ToPublic())
+	if err != nil {
+		return fmt.Errorf("failed to marshal create speech request: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, a.speechURL, bytes.NewReader(data))
+	if err != nil {
+		return fmt.Errorf("failed to create speech request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	if a.apiKey != "" {
+		req.Header.Set("Authorization", "Bearer "+a.apiKey)
+	}
+
+	sr := new(db.CreateSpeechResponse)
+	code, err := sendSpeechRequest(a.client, req, sr)
+	if err != nil {
+		l.Error("failed to send speech create request", "err", err)
+		sr.Error = z.Pointer(err.Error())
+	}
+
+	sr.StatusCode = code
+	sr.RequestID = speechRequest.ID
+	sr.Done = true
+
+	if err := gdb.Transaction(func(tx *gorm.DB) error {
+		if err := db.Create(tx, sr); err != nil {
+			return err
+		}
+
+		return tx.Model(speechRequest).Where("id = ?", speechRequest.ID).Update("done", true).Error
+	}); err != nil {
+		l.Error("failed to store speech create response", "err", err)
+	}
+
+	return nil
+}
+
+func sendSpeechRequest(client *http.Client, req *http.Request, respObj *db.CreateSpeechResponse) (code int, err error) {
+	var res *http.Response
+	res, err = client.Do(req)
+	if err != nil {
+		return 0, err
+	}
+
+	defer func() {
+		err = errors.Join(err, res.Body.Close())
+	}()
+
+	code = res.StatusCode
+	if code < http.StatusOK || code >= http.StatusBadRequest {
+		return code, cclient.DecodeError(res)
+	}
+
+	respObj.Content, err = io.ReadAll(res.Body)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	return code, nil
+}
```

Comment:
Can this be moved to the `client` package? I'd rather keep the `DecodeError` function unexported.

Reply:
done.